### ISSUE TITLE
Installing servers for testing

### ISSUE THREAD

**Issue by dmill-bz** [spider/spider#35]
Feel free to add orient to this.
###### Download
- [Gremlin-server](https://www.apache.org/dist/incubator/tinkerpop/3.0.0-incubating/apache-gremlin-server-3.0.0-incubating-bin.zip)
- [gremlin-server-php.yaml](https://raw.githubusercontent.com/PommeVerte/gremlin-php/master/src/tests/gremlin-server-php.yaml)
- [gremlin-php-script.groovy](https://raw.githubusercontent.com/PommeVerte/gremlin-php/master/src/tests/gremlin-php-script.groovy)
- [neo4j-empty.properties](https://raw.githubusercontent.com/PommeVerte/gremlin-php/master/src/tests/neo4j-empty.properties)
###### Installation

Extract Gremlin-server and go into the created folder.

Neo4J is required for the full test suit (it serves as the transaction enabled graph). It is not bundled with gremlin-server by default so you will need to manually install it with:

``` bash
bin/gremlin-server.sh -i org.apache.tinkerpop neo4j-gremlin 3.0.0-incubating
```

Copy the following files :

``` bash
cp path/to/gremlin-server-php.yaml <gremlin-server-root-dir>/conf/
cp path/to/neo4j-empty.properties <gremlin-server-root-dir>/conf/
cp path/to/gremlin-php-script.groovy <gremlin-server-root-dir>/scripts/
```

You will then need to run gremlin-server in the following manner (it's important to run it from the `<gremlin-server-root-dir>`):

``` bash
bin/gremlin-server.sh conf/gremlin-server-php.yaml
```
###### Download
- [Neo4J Server](http://neo4j.com/download/) (Community version is tested)
###### Installation

Once the server is up and running, connect to [http://localhost:7474/browser](http://localhost:7474/browser) 
- Change the username:password from `neo4j:neo4j` to `neo4j:j4oen`
- Click on the star tab and import [this script](https://gist.githubusercontent.com/PommeVerte/46ed5dee72688ee2faf9/raw/12804b280e3e9361319afb5a4bd2084505aed1d0/Neo4J%2520modern%2520graph) that will create a "modern" graph.
###### Download
- [OrientDB Community](http://orientdb.com/download/) (Community version is tested)
###### Installation

Simply extract the directory and place it anywhere you like.

That's really it. The server can be accessed by connecting to : [http://localhost:2480/](http://localhost:2480/).

Spider tests use a default graph that is bundled with Orient called "GratefulDeadConcerts". You do not need to do anything to prepare this graph. The tests also use the default username and password: "root" and "root". If you change these for the GratefulDeadConcerts graph, be sure to change the credentials in `tests/Unit/Drivers/OrientDB/DriverTest.php`

You should be good to go.

**Comment by dmill-bz** [comment]
Lets also use this issue as a means to discuss the CI configuration. I can provide the Travis configuration for getting gremlin-server up and running.

**Comment by electricjones** [comment]
I was not aware that Travis allowed for setup of other database systems. Learn something new everyday.

**Comment by electricjones** [comment]
This may also be a good place for an early discussion about Database Fixtures for testing, though a full implementation may need to wait until a Schema Builder is on its way.

**Comment by electricjones** [comment]
So, I got Neo4j up and running (its nice to see tests pass), but Gremlin is giving me an error when I try to install Neo4j:
`Could not find or load main class org.apache.tinkerpop.gremlin.server.util.GremlinServerInstall`

I `cd` into the bin directory and it installs, but says that its detected a non-standard directory. Then, when I try to run from the root directory I get the same error.

It appears that I am not able to run the install script from the root directory? I am on a windows machine if that makes any difference.

**Comment by electricjones** [comment]
Oops, just noticed the bat file. Don't know what I was thinking.

**Comment by electricjones** [comment]
Well, I got gremlin-server up and running, but the test fail with errors:

`ServerExeption: The script submitted for processing evaluated in the ScriptEngine with errors and could not be processed. Check the script submitted for syntax errors or other problems and resubmit. :   No such property: g for class Script2 >`

**Comment by dmill-bz** [comment]
ohhh sorry yeah the instructions were linux based. Didn't think about that.

And that's strange because they all seem to pass for me. If you run `bin/gremlin.bat -v start` do you get any warnings?

**Comment by dmill-bz** [comment]
Great orient tests pass on my end. Now I'll be able to tell if I break anything.

**Comment by electricjones** [comment]
I don't have a gremlin.bat, but a gremlin-server.bat I get an error:
`GremlinServer - Configuration file at -v could not be found or parsed properly.`

Glad to hear the orient tests pass.

**Comment by dmill-bz** [comment]
Are you running the command from the root folder (it's a little odd but this is required atm)? And is `gremlin-server-php.yaml` in the right place with the proper rights?

**Comment by dmill-bz** [comment]
lets try this again I'm mixing titan and gs. 
`bin/gremlin-server.bat -v conf/gremlin-server-php.yaml`

**Comment by electricjones** [comment]
`gremlin-server-php.yaml` is at `conf/gremlin-server-php.yaml`.

I am running it from the folder I extracted with `bin',`conf`,`data`,`ext`, etc.

I get the same error when I run `bin/gremlin-server.bat -v conf/gremlin-server-php.yaml`: "Configuration file at -v could not be found or parsed correctly. 

When I try `bin\gremlin-server.bat conf\gremlin-server-php.yaml` It seems to start correctly, but my tests fail.

At least it seems it's not a problem with spider, but the server installation. I used the link provided to download gremlin. Could we be using different versions?

**Comment by dmill-bz** [comment]
That's odd, yes we're using the same version. Does `bin/gremlin-server.bat conf/gremlin-server-php.yaml -v` work?

Do you have any output in the `log` dir? From the error you got during the tests it looks like the traversal object fails to initialize correctly.

**Comment by electricjones** [comment]
Yes, placing -v at the end works. Where is the `log` directory?

**Comment by dmill-bz** [comment]
ok do you get any warnings when you run the server? from all the output? The logs probably need turning on, I'll have to look into how it's done

**Comment by dmill-bz** [comment]
Oh I just thought of something. Do you have java 8 on your system?

**Comment by electricjones** [comment]
I do have Java SDK 8. It's also required for Orient.

**Comment by electricjones** [comment]
So, I looked through my output and I may not have groovy installed. I do get this warning "Could not initialize gremlin-groovy script engine". and a couple others. I'll post them later tonight

**Comment by dmill-bz** [comment]
ok I'll try and help with those

**Comment by electricjones** [comment]
I also get a warning `Graph [graphT] configured at [conf/neo4j-empty.properties could not be instantiated and will not be available in Gremlin Server.`

**Comment by electricjones** [comment]
I think I've tracked down the problem, but I don't know how to fix it. The neo4j dependency is not installing correctly.

![gremlinerror](https://cloud.githubusercontent.com/assets/5412413/9153467/cad6e662-3e1d-11e5-8c10-c9663b0af0d8.JPG)

This sounds like an issue that needs be be opened with tinkerpop, but I'm not sure where.

**Comment by dmill-bz** [comment]
Ahh indeed that does look like a problem. I'm surprised it hasn't been reported before. The easiest way is either to write an email to the [gremlin-users group](https://groups.google.com/forum/#!forum/gremlin-users) or directly in their [JIRA](https://issues.apache.org/jira/browse/TINKERPOP3/?selectedTab=com.atlassian.jira.jira-projects-plugin:issues-panel).
They also have a dev mailing list but it might be a pain to signup and get flooded with emails just to report an issue. 

With that said that means only transaction tests should fail for you.

**Comment by dmill-bz** [comment]
oh and while I'm at it. I've found that the neo4J dependency sometimes fails. When it does you need to delete the folder and re-extract the zip then try again or it will complain that a dependency with that name already exists. 
(though from the look of things, that's what you've already been doing)

**Comment by electricjones** [comment]
Well, I'm getting 4 errors and 3 failures, but it seems like they are all because of the same `ServerException` and problem with the `ScriptEngine`, however, I don't get any warnings when using any of the standard gremlin-server yamls, and some of them use the gremlin-groovy script engine. I won't pretend to understand how the system holds together after an hour looking through the code, but it seems that my neo4j extension is half-installing, and when I start gremlin-server it for some reason mishandles the gremlin-groovy ScriptEngine so it doesn't load correctly.

**Comment by dmill-bz** [comment]
Well it looks like the problem with loading neo4j comes from the fact that additional [dependencies don't seem to get the same OS based normalizing](https://github.com/apache/incubator-tinkerpop/blob/master/gremlin-groovy/src/main/groovy/org/apache/tinkerpop/gremlin/groovy/util/DependencyGrabber.groovy#L92-L101) of paths as the original set of dependencies (though I'm not 100% positive)
Think reporting an issue is in order

**Comment by electricjones** [comment]
I have just posted an issue to their JIRA. 

https://issues.apache.org/jira/browse/TINKERPOP3-804?jql=project%20%3D%20TINKERPOP3%20AND%20resolution%20%3D%20Unresolved%20AND%20priority%20%3D%20Major%20ORDER%20BY%20key%20DESC

For now, I'm just going to skip those tests and rely on you to make sure they don't break :smile:.

Can we set these tests up on Travis?

**Comment by dmill-bz** [comment]
Yes. It's a bit of a mess but here's the gremlin-server Travis setup (we can work into getting this into a script or something):

``` yaml
before_install:
  - sudo apt-get update > /dev/null
install:
  # Composer install for Spider. This is a requirement as we need gremlin-php conf files
  - composer install 
  # Install Java 8
  - sudo apt-get install -y python-software-properties
  - sudo add-apt-repository -y ppa:webupd8team/java
  - sudo apt-get update
  - sudo apt-get install -y oracle-java8-installer
  - sudo update-alternatives --auto java
  - sudo update-alternatives --auto javac
  - export JAVA_HOME=/usr/lib/jvm/java-8-oracle
  # Install gremlin-server
  - wget --no-check-certificate -O $HOME/apache-gremlin-server-3.0.0-incubating-bin.zip https://www.apache.org/dist/incubator/tinkerpop/3.0.0-incubating/apache-gremlin-server-3.0.0-incubating-bin.zip
  - unzip $HOME/apache-gremlin-server-3.0.0-incubating-bin.zip -d $HOME/
  # get gremlin-server configuration files  
  - cp vendor/brightzone/gremlin-php/src/tests/gremlin-php-script.groovy $HOME/apache-gremlin-server-3.0.0-incubating/scripts/
  - cp vendor/brightzone/gremlin-php/src/tests/gremlin-server-php.yaml $HOME/apache-gremlin-server-3.0.0-incubating/conf/
  - cp vendor/brightzone/gremlin-php/src/tests/neo4j-empty.properties $HOME/apache-gremlin-server-3.0.0-incubating/scripts/
  # get neo4j dependencies
  - cd $HOME/apache-gremlin-server-3.0.0-incubating
  - bin/gremlin-server.sh -i org.apache.tinkerpop neo4j-gremlin 3.0.0-incubating
  # Start gremlin-server in the background and wait for it to be available
  - bin/gremlin-server.sh conf/gremlin-server-php.yaml > /dev/null 2>&1 &
  - cd $TRAVIS_BUILD_DIR
  - sleep 60
```

It should work out of the box (or with some minor tweaks).
There's probably a better way of doing this, though I'm not sure the free Travis accounts can hold onto and reuse artifacts. With that said this takes a bit of time but it's nothing outrageous.

**Comment by dmill-bz** [comment]
Also I think travis supports Neo4J via : 

``` yaml
services:
  - neo4j
```

I'm not sure what login/password to expect though. neither how to set up the "modern" graph. (I guess this is where fixtures would be useful)

**Comment by dmill-bz** [comment]
And here's some reading material for orient in travis : http://odino.org/bootstrap-of-a-travis-ci-test-machine/

**Comment by electricjones** [comment]
I will take a look though all of this next week and see if I can get something setup. As far as fixtures go, the ideal would be to use a setup procedure to to build up a graph we want. Truthfully, the modern graph is simple enough, that I could replicate it for Orient an then all three would use the same graph.

Then, as the test suite begins, we can build that graph up from scratch and destroy it after tests are complete. I'm sure there is a better way, but its an option.

**Comment by electricjones** [comment]
For now, I'm gonna let tinkerpop take a look at that issue and move onto adding to the documentation and working through the Top Level Spider API in the direction of 0.4.

**Comment by dmill-bz** [comment]
Ok. I'm fixing another TP issue (tree one) after that I'll give the CypherProcessor a look. (If I find the time I'll try correcting the issue you found)

**Comment by electricjones** [comment]
It seems like a fix has been pushed to master, but now I'm having problems building from the master branch, lol. My machine needs a good scrubbing (and to be linux, but its work. Whatcha gonna do?).

**Comment by dmill-bz** [comment]
Yeah it's not functional atm anyways. 
I don't really know how your env is set up but [this link](http://www.mkyong.com/maven/how-to-install-maven-in-windows/) is good for installing maven.
As for building from master, `mvn clean -DskipTests=true install` should do the trick.
You can find the zip in `gremlin-server\target\apache-blablabla.zip`

**Comment by dmill-bz** [comment]
Can we set travis up for spider/spider ? Or should I run the tests in a fork? I would like to make a PR with the databases set up so we can remove the skipped tests.

**Comment by dmill-bz** [comment]
(actually I'll do it in a fork. That way I'll have more control over stopping the tests manually as I go.)

**Comment by electricjones** [comment]
I was actually just about to open an issue about this. As the light for 0.3 is at the end of the tunnel, we should make sure our building, coverage, and analysis is going.

For now, a fork would be best.

**Comment by electricjones** [comment]
Yeah, I'm getting sick of all those skipped tests, too, lol.

Also, I've seen some headway on my JIRA issue. Is it possible for me to get a build with those changes?

**Comment by dmill-bz** [comment]
I've been meaning to look into what Stephen was asking. I'll try to get around to that tomorrow to see if the fix works (fingers crossed)

**Comment by dmill-bz** [comment]
ok So Stephen's fix works. Which means that if you can build the `tp30` branch you should be good to go. Let me know what you're missing for this to work, maybe I can help.
- [install maven](http://www.mkyong.com/maven/how-to-install-maven-in-windows/) (you might also need to add `JAVA_HOME` to `PATH`)
- `git clone --branch tp30 https://github.com/apache/incubator-tinkerpop.git`
- `cd incubator-tinkerpop`
- `mvn clean -DskipTests=true install`
- Then you will find the zips you need in `gremlin-server\target\` and `gremlin-console\target\`. The console is for if you want to access the db manually (not mandatory for spider testing)
- You might need to change the configuration files linked above to suit the new `3.0.1` formats. Let me know once you get here.

**Comment by electricjones** [comment]
Alright, snag number 2 (or 3). As I tried to build gremlin-server from the master branch, it failed. It seems my java version is 1.8.0_51 and tinkerpop only accepts to to version 1.8.0_40.

**Comment by dmill-bz** [comment]
I've been using 1.8.0_51 as well. What error messages are you getting?

**Comment by electricjones** [comment]
![gremlinerror](https://cloud.githubusercontent.com/assets/5412413/9364112/4f357e28-4670-11e5-90c6-ed04896f5e7d.JPG)

**Comment by dmill-bz** [comment]
That looks like the neo4J installation error (mixup?).

**Comment by electricjones** [comment]
My bad:

![tinkerpop-error](https://cloud.githubusercontent.com/assets/5412413/9364200/c62ab30e-4670-11e5-8ea7-f905a24ea5d1.JPG)

**Comment by dmill-bz** [comment]
ahh that's odd. it seems to say you have java `1.8.0_5` what does `java -version` return? If it's `1.8.0_51` then perhaps your `JAVA_HOME` is pointing to the wrong version or your `PATH` is incorrect. 

Also, is this the `master` or `tp30` branch? You'll want the `tp30`.

**Comment by dmill-bz** [comment]
Here are my env variables for reference:
- `JAVA_HOME` : `C:\Program Files\Java\jdk1.8.0_51`
- `M2_HOME` : `C:\Program Files\maven`
- `MAVEN_HOME` : `C:\Program Files\maven`
- `Path` (`M2_HOME` and `JAVA_HOME` sections are relevant): 
  
  ```
  C:\Program Files\Parallels\Parallels Tools\Applications;%SystemRoot%\system32;%SystemRoot%;%SystemRoot%\System32\Wbem;%SYSTEMROOT%\System32\WindowsPowerShell\v1.0\;%M2_HOME%\bin;%JAVA_HOME%\bin
  ```

**Comment by electricjones** [comment]
Alright, I finally got gremlin-server installed, and now when I start it, I get an error that it can't initialize graphT. Do I need to modify the conf files at all?

**Comment by dmill-bz** [comment]
Yeah you might. I'll give it a run to see what needs to be changed for 3.0.1

**Comment by electricjones** [comment]
Also, it just occurred to me that the neo plugin may not have installed. I deleted all my previous binaries and builds, extracted the new build, and ran the command to install the neo bridge, but it said that it could not install because that plugin already existed. I assumed that (since this is a new build) that plugin was included by default, but that may not be the case.

Does it store plugins in some other classpath I need to clean? Or is there a strange caching issue here?

**Comment by dmill-bz** [comment]
Ok so it's weird that the only error you get is in regards to graphT. (which is neo)
It's also weird that it would install neo by default (it shouldn't). You've done this already but lets try again. Try deleting the gremlin-server folder and re-extracting `gremlin-server/target/apache-blabla.zip` file (to your Downloads folder for example). Then running the whole neo command again.
If you get the error again then we might have another issue to figure out.

This wouldn't be related to any change in configuration as graphT should be initialized by then.

**Comment by electricjones** [comment]
!Qapla!

It must have been a caching issue. I extracted to an entirely new location, followed the directions at the start (switching for windows) and everything installed, started up, and my gremlin tests pass!

We can probably close this issue now, and add this to the documentation. Thank you for all your help on this.

PS: I replaced the neoproperties that the build provided. Also, should we add that for windows users, they must build server on their own until the next release?
